### PR TITLE
Add getsockopt/setsockopt error checking and convenience methods.

### PR DESCRIPTION
**Pull request by technorama**

### CODE REVIEW

File: src/socket/socket.cr [hunk 1]
```diff
@@ -57,6 +57,52 @@ class Socket < FileDescriptorIO
...
+  end
+
+  def recv_buf_size
+    getsockopt LibC::SO_RCVBUF, 0
+  end
+
+  def recv_buf_size= val : Int32
+    setsockopt LibC::SO_RCVBUF, val
+    val
+  end
```
Comment (jhass) [inline src/socket/socket.cr]: I think I'd prefer `buffer` to be written out here.

File: src/socket/socket.cr [hunk 2]
```diff
@@ -57,6 +57,52 @@ class Socket < FileDescriptorIO
...
+  def reuse_address= val : Bool
+    v = val ? 1 : 0
+    setsockopt LibC::SO_REUSEADDR, v
+    val
+  end
+
+  # optval is restricted to Int32 until sizeof works on variables
+  # returns the modified optval
+  def getsockopt optname, optval : Int32, level = LibC::SOL_SOCKET
+    optsize = LibC::SocklenT.cast(sizeof(Int32))
```
Comment (jhass) [inline src/socket/socket.cr]: As @asterite mentioned on IRC you can `sizeof(typeof(optval))`

File: src/socket/socket.cr [hunk 3]
```diff
@@ -57,6 +57,51 @@ class Socket < FileDescriptorIO
...
+  # returns the modified optval
+  def getsockopt optname, optval, level = LibC::SOL_SOCKET
+    optsize = LibC::SocklenT.cast(sizeof(typeof(optval)))
+    ret = LibC.getsockopt(fd, level, optname, (pointerof(optval) as Void*), out optsize)
+    raise Errno.new("getsockopt") if ret == -1
+    optval
+  end
+
+  # optval is restricted to Int32 until sizeof works on variables
+  def setsockopt optname, optval : Int32, level = LibC::SOL_SOCKET
```
Comment (jhass) [inline src/socket/socket.cr]: Missing updates? ;) Also I think a spec that calls this with different types for optval would be awesome.

**Comment by technorama** [comment]
@jhass all comments implemented except non-int32 specs.  Almost all socket options take an int.  There are a few protocol or platform specific exotics that use something else.  I'm not going to implement what I'm not familiar with.

PR's welcome.

**Comment by jhass** [comment]
I'm afraid it's a rather critical spec though, since if you pass different types, `optval` becomes a union, changing behaviour of `sizeof(typeof(optval))`

**Pull request merged**